### PR TITLE
Improve command not found errors

### PR DESCRIPTION
**Pull request by LnL7**
Improves error message when git/mercurial are not available. eg.

```
error: program 'git' failed with exit code 127
The program 'git' is currently not installed. It is required for builtins.fetchGit
You can install it by running the following:
nix-env -f '<nixpkgs>' -iA git
```

### CODE REVIEW

File: src/libexpr/primops/fetchGit.cc [hunk 1]
```diff
@@ -223,17 +223,29 @@ static void prim_fetchGit(EvalState & state, const Pos & pos, Value * * args, Va
...
+        v.attrs->sort();
+
+        if (state.allowedPaths)
+            state.allowedPaths->insert(gitInfo.storePath);
+    } catch (ExecError & e) {
+        if (e.status / 256 == 127) {
+            std::ostringstream out;
+            out << e.what() << std::endl;
+            out << "The program 'git' is currently not installed. It is required by builtins.fetchGit" << std::endl;
+            out << "You can install it by typing one of the following:" << std::endl;
```
Comment (grahamc) [inline src/libexpr/primops/fetchGit.cc]:
```suggestion
            out << "You can install it by running the following:" << std::endl;
```

File: src/libexpr/primops/fetchMercurial.cc [hunk 2]
```diff
@@ -188,18 +188,30 @@ static void prim_fetchMercurial(EvalState & state, const Pos & pos, Value * * ar
...
+        v.attrs->sort();
+
+        if (state.allowedPaths)
+            state.allowedPaths->insert(hgInfo.storePath);
+    } catch (ExecError & e) {
+        if (e.status / 256 == 127) {
+            std::ostringstream out;
+            out << e.what() << std::endl;
+            out << "The program 'hg' is currently not installed. It is required by builtins.fetchMercurial" << std::endl;
+            out << "You can install it by typing one of the following:" << std::endl;
```
Comment (grahamc) [inline src/libexpr/primops/fetchMercurial.cc]:
```suggestion
            out << "You can install it by running the following:" << std::endl;
```

**Comment by grahamc** [comment]
This is awesome!

File: src/libexpr/primops/fetchGit.cc [hunk 3]
```diff
@@ -223,17 +223,29 @@ static void prim_fetchGit(EvalState & state, const Pos & pos, Value * * args, Va
...
+        mkString(*state.allocAttr(v, state.sOutPath), gitInfo.storePath, PathSet({gitInfo.storePath}));
+        mkString(*state.allocAttr(v, state.symbols.create("rev")), gitInfo.rev);
+        mkString(*state.allocAttr(v, state.symbols.create("shortRev")), gitInfo.shortRev);
+        mkInt(*state.allocAttr(v, state.symbols.create("revCount")), gitInfo.revCount);
+        v.attrs->sort();
+
+        if (state.allowedPaths)
+            state.allowedPaths->insert(gitInfo.storePath);
+    } catch (ExecError & e) {
+        if (e.status / 256 == 127) {
```
Comment (edolstra) [inline src/libexpr/primops/fetchGit.cc]: Should be `WEXITSTATUS(e.status) == 127`.

File: src/libexpr/primops/fetchGit.cc [hunk 4]
```diff
@@ -223,17 +223,29 @@ static void prim_fetchGit(EvalState & state, const Pos & pos, Value * * args, Va
...
+            std::ostringstream out;
+            out << e.what() << std::endl;
+            out << "The program 'git' is currently not installed. It is required for builtins.fetchGit" << std::endl;
+            out << "You can install it by running the following:" << std::endl;
+            out << "nix-env -f '<nixpkgs>' -iA git";
+            throw ExecError(e.status, out.str());
```
Comment (edolstra) [inline src/libexpr/primops/fetchGit.cc]: This should use format strings:
```
throw ExecError("%s\nThe program 'git' is currently not installed. ...", e.what());
```

**Comment by LnL7** [comment]
ping

**Comment by grahamc** [comment]
Just built the rebased version, and checked out the code at https://github.com/NixOS/nix/pull/2498/files?w=1 

Looks good to me I think

File: src/libexpr/primops/fetchGit.cc [hunk 5]
```diff
@@ -225,17 +225,28 @@ static void prim_fetchGit(EvalState & state, const Pos & pos, Value * * args, Va
...
+        state.mkAttrs(v, 8);
+        mkString(*state.allocAttr(v, state.sOutPath), gitInfo.storePath, PathSet({gitInfo.storePath}));
+        mkString(*state.allocAttr(v, state.symbols.create("rev")), gitInfo.rev);
+        mkString(*state.allocAttr(v, state.symbols.create("shortRev")), gitInfo.shortRev);
+        mkInt(*state.allocAttr(v, state.symbols.create("revCount")), gitInfo.revCount);
+        v.attrs->sort();
+
+        if (state.allowedPaths)
+            state.allowedPaths->insert(state.store->toRealPath(gitInfo.storePath));
+    } catch (ExecError & e) {
```
Comment (edolstra) [inline src/libexpr/primops/fetchGit.cc]: It's better to put this check in `exportGit` because (on the flakes branch) it's used in other places as well.

Comment (LnL7) [reply src/libexpr/primops/fetchGit.cc]: Sure, should I do the same for mercurial?

**Comment by grahamc** [comment]
> Sure, should I do the same for mercurial?

Yeah, probably should :)

**Comment by roberth** [comment]
Still relevant but needs a rebase. Current output with flakes is:

```
error: error: --- SysError -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- nix-build
executing 'git': No such file or directory
error: --- Error ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- nix-build
program 'git' failed with exit code 1
```

**Comment by LnL7** [comment]
I'm happy to rebase this, but it's been open for quite some time without getting attention.

**Pull request closed without merging**